### PR TITLE
Created initial function for IMD bar chart

### LINKED ISSUE #33: Shared chart function - IMD decile
Create a function to create a chart based on supplied data.

This will produce a highcharts chart object with the majority of the formatting taken care of which can subsequently be tweaked by the user if required.

Inputs: dataframe, reference for columns for IMD decile and value to report, chart title and axis labels
Outputs: chart object
Optional parameters: whether to highlight core20 population (deciles 1 and 2)

### PR DESCRIPTION
**Pull request by nhsbsa-stbuc**

### CODE REVIEW

File: R/theme_nhsbsa.R [hunk 1]
```diff
@@ -177,3 +177,45 @@ theme_nhsbsa_hc <- function(family = "TT Arial") {
...
+#'
+#' @param palette Which colour palette to use from the `nhsbsaR` package.
+#' @param stack Stack option for highcharter.
+#'
+#' @return Theme data
+#' @export
+theme_nhsbsa <- function(hc, palette = NA, stack = "normal") {
+
+  # Load theme from nhsbsaR package
+  theme_nhsbsa_hc <- nhsbsaR::theme_nhsbsa_hc(family = "Frutiger W01")
```
Comment (nhsbsa-mamcp) [inline R/theme_nhsbsa.R]: Don't need to namespace functions within same package

File: R/theme_nhsbsa.R [hunk 2]
```diff
@@ -177,3 +177,45 @@ theme_nhsbsa_hc <- function(family = "TT Arial") {
...
+
+  # Add the plot options
+  theme_nhsbsa_hc$plotOptions <- list(
+    series = list(stacking = stack, borderWidth = 0),
+    bar = list(groupPadding = 0.1)
+  )
+
+  # Add the palettes (hack the highlight palette to have a lighter grey)
+  theme_nhsbsa_hc$colors <- nhsbsaR::palette_nhsbsa(palette = palette)
```
Comment (nhsbsa-mamcp) [inline R/theme_nhsbsa.R]: Don't need to namespace functions within same package

File: R/theme_nhsbsa.R [hunk 3]
```diff
@@ -177,3 +177,45 @@ theme_nhsbsa_hc <- function(family = "TT Arial") {
...
+  theme_nhsbsa_hc$colors[theme_nhsbsa_hc$colors == "#768692"] <- "#d1d5d6"
+    theme_nhsbsa_hc$colAxis <- list(
+      min = 0,
+      minColor = nhsbsaR::palette_nhsbsa(palette = "gradient")[1],
+      maxColor = nhsbsaR::palette_nhsbsa(palette = "gradient")[2]
```
Comment (nhsbsa-mamcp) [inline R/theme_nhsbsa.R]: Don't need to namespace functions within same package

File: R/theme_nhsbsa.R [hunk 4]
```diff
@@ -177,3 +177,45 @@ theme_nhsbsa_hc <- function(family = "TT Arial") {
     )
   )
 }
+
+
+#' Custom NHSBSA highcharter theme
```
Comment (nhsbsa-mamcp) [inline R/theme_nhsbsa.R]: Is there a need to have 2 different hc themes, could this one replace the existing? Or are you just adding for review to then replace existing?

File: R/theme_nhsbsa.R [hunk 5]
```diff
@@ -177,3 +177,45 @@ theme_nhsbsa_hc <- function(family = "TT Arial") {
...
+      minColor = nhsbsaR::palette_nhsbsa(palette = "gradient")[1],
+      maxColor = nhsbsaR::palette_nhsbsa(palette = "gradient")[2]
+    )
+
+    # Style based on the NHS frontend toolkit
+    theme_nhsbsa_hc$xAxis$className <- "nhsuk-body-s"
+    theme_nhsbsa_hc$yAxis$className <- "nhsuk-body-s"
+
+    # Add the theme to the chart and then remove the credits afterwards (currently
+    # does not work to do this within the theme)
```
Comment (nhsbsa-mamcp) [inline R/theme_nhsbsa.R]: Did you remove the code for removing credits? I don't see it.

File: R/theme_nhsbsa.R [hunk 6]
```diff
@@ -177,3 +177,45 @@ theme_nhsbsa_hc <- function(family = "TT Arial") {
...
+    hc |>
+      highcharter::hc_add_theme(hc_thm = theme_nhsbsa_hc)
+    #  |>     highcharter::hc_colors(colors = nhsbsaR::palette_nhsbsa(palette = palette))
```
Comment (nhsbsa-mamcp) [inline R/theme_nhsbsa.R]: Are you not sure about the commented line? Or is this no longer needed following the above palette code?

File: R/chart_hc_imd_bar.R [hunk 7]
```diff
@@ -0,0 +1,220 @@
+#' chart_hc_imd_bar function
```
Comment (nhsbsa-mamcp) [inline R/chart_hc_imd_bar.R]: Could be more descriptive, like "Highcharts bar plot of IMD deciles"

File: R/chart_hc_imd_bar.R [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    highlight_core20 = FALSE,
+    show_lbl = FALSE,
+    value_dp = 0,
+    value_prefix = "",
+    order_of_magnitude = ""
+) {
+
+# Input review --------------------------------------------------------------------------------
+
+  # review input parameters, providing error/warning messages as required and aborting call
```
Comment (nhsbsa-mamcp) [inline R/chart_hc_imd_bar.R:68]: Could validate mandatory args also

File: R/chart_hc_imd_bar.R [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+
+# Data object ---------------------------------------------------------------------------------
+
+  # define the data object based on parameters
+  df <- df |>
+    dplyr::mutate(IMD_COL = df[[imd_col]],
+                  VAL_COL = df[[value_col]])
+
```
Comment (nhsbsa-mamcp) [inline R/chart_hc_imd_bar.R:114]: Use `rename` to avoid duplicating columns.

File: R/chart_hc_imd_bar.R [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+  # adjust values based on supplied order of magnitude
+  if(order_of_magnitude != ""){
+    df <- df |>
+      dplyr::mutate(VAL_COL = VAL_COL / switch(order_of_magnitude,
+                                               "k" = 1000,
+                                               "m" = 1000000,
+                                               "bn" = 1000000000,
+                                               "tn" = 1000000000000)
+                    )
+  }
```
Comment (nhsbsa-mamcp) [inline R/chart_hc_imd_bar.R:135]: Alternative way:

change arg to `order_of_magnitude = "identity"`

```
magnitudes <- c(identity = 10^0, k = 10^3, m = 10^6, bn = 10^9, tn = 10^12)

df <- df |>
  dplyr::mutate(VAL_COL = VAL_COL / magnitudes[[order_of_magnitude]])
```

Comment (nhsbsa-mamcp) [reply R/chart_hc_imd_bar.R:135]: I see you use the fact that order_of_magnitude is `""` by default later, so would have to add below after. That is a bit icky, so maybe not use my suggestion!

`order_of_magnitude <- ifelse(order_of_magnitude = "identity", "", order_of_magnitude)`

File: R/chart_hc_imd_bar.R [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    )
+  }
+
+  # add IMD decile and value to the tooltip
+  tooltip_text <- paste0(
+    tooltip_text,
+    "<b>IMD decile:</b> {point.IMD_COL} <br>",
+    "<b>",metric_def,":</b> ",value_prefix,"{point.VAL_COL:,.",value_dp,"f}",order_of_magnitude
+  )
+
```
Comment (nhsbsa-mamcp) [inline R/chart_hc_imd_bar.R:155]: Using `[{glue}](https://glue.tidyverse.org/)` allows more readable syntax:

```
tooltip_text <- ""

if(highlight_core20){
  tooltip_text <- "<b>Core20 Classification:</b> {point.GRP_COL}<br>"
}

# add IMD decile and value to the tooltip
tooltip_text <- glue(
  "{tooltip_text}<b>IMD decile:</b> {{point.IMD_COL}}<br>\\
   <b>{metric_def}:</b> {value_prefix}{{point.VAL_COL:,.{value_dp},f}}\\
   {order_of_magnitude}"
)
```

File: R/chart_hc_imd_bar.R [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+                        inside = FALSE,
+                        style = list(color = "#000000")
+                        )
+    ) |>
+    theme_nhsbsa_hc() |>
+    highcharter::hc_yAxis(title = list(text = y_axis_lbl)) |>
+    highcharter::hc_xAxis(
+      min = 1,
+      max = 11, # Pad to ensure we can see the 10 label
+      # categories = c(NA, "1<br>Most<br>deprived", rep(NA, 8), "10<br>Least<br>deprived"),
```
Comment (nhsbsa-mamcp) [inline R/chart_hc_imd_bar.R]: Commented line, not sure about this, or left in by mistake?

File: R/chart_hc_imd_bar.R [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+      categories = c(NA, "1<br>Most<br>deprived", seq(2,9), "10<br>Least<br>deprived",""),
```
Comment (nhsbsa-mamcp) [inline R/chart_hc_imd_bar.R]: What does the NA at beginning do? Never seen this used before.

File: R/chart_hc_imd_bar.R [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    highcharter::hc_credits(enabled = TRUE)  |>
+    highcharter::hc_tooltip(
+      headerFormat = "",
+      pointFormat = tooltip_text
+    ) |>
+    highcharter::hc_legend(enabled = TRUE)
+
+
+  # specific formatting when highlighting core20
+  if(highlight_core20 == TRUE){
```
Comment (nhsbsa-mamcp) [inline R/chart_hc_imd_bar.R:207]: Don't need the `== TRUE`:

`if(highlight_core20){`

File: R/chart_hc_imd_bar.R [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+
+
+# Tooltip creation ----------------------------------------------------------------------------
+
+  # create the tooltip object
+  tooltip_text <- ""
+
+
+  # if core20 should be highlighted add this to the tooltip
+  if(highlight_core20 == TRUE){
```
Comment (nhsbsa-mamcp) [inline R/chart_hc_imd_bar.R:145]: Don't need the `== TRUE`:

`if(highlight_core20){`

File: R/chart_hc_imd_bar.R [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+          )
+      )
+    )
+  }
+
+
+# Return output -------------------------------------------------------------------------------
+
+  # return the plot object
+  return(plt)
```
Comment (nhsbsa-mamcp) [inline R/chart_hc_imd_bar.R:225]: Don't need the return - functions always return the last object

Comment (nhsbsa-mamcp) [reply R/chart_hc_imd_bar.R:225]: Also, if you want to suppress it can use `invisible(something)`

**Review by nhsbsa-mamcp — APPROVED** [review]
Hope you don't think I am nitpicking at all - all just suggestions, it all works well as is.

File: R/chart_hc_imd_bar.R (hunk 13 above)
Comment (nhsbsa-stbuc) [reply R/chart_hc_imd_bar.R]: Think this is the point where the axis cross so leaving this NA starts the labels at 1 for the first bar.  If omitted the first bar is labelled 2

**Comment by nhsbsa-stbuc** [comment]
Nitpicking is all good.  Made some changes based on suggestions.  Backed out changes to the theme as there is a dedicated PR for those now and will merge that before this one to make sure this function references the new theme function.

**Pull request merged**